llama.cpp

```diff
@@ -1828,19 +1826,149 @@ struct llama_model_loader {
         }
     }
 
-    std::string get_arch_name() const {
-        const auto kv = LLM_KV(LLM_ARCH_UNKNOWN);
+    private:
+    template <typename T> struct gk_get_arrlen  { T & output; };
+    template <typename TI, typename TO> struct gk_set_literal { TI & input; TO & output; };
+    template <typename T>
+    void gk_set(int kid, T & result) {
+        (void)result;
+        throw std::runtime_error(format("request for key id %d with unhandled result type: %s", kid, typeid(T).name()));
+    }
+
+    void gk_set(int k, uint8_t     & r) { r = gguf_get_val_u8  (ctx_gguf, k); }
+    void gk_set(int k, uint16_t    & r) { r = gguf_get_val_u16 (ctx_gguf, k); }
+    void gk_set(int k, uint32_t    & r) { r = gguf_get_val_u32 (ctx_gguf, k); }
+    void gk_set(int k, uint64_t    & r) { r = gguf_get_val_u64 (ctx_gguf, k); }
+    void gk_set(int k, int8_t      & r) { r = gguf_get_val_i8  (ctx_gguf, k); }
+    void gk_set(int k, int16_t     & r) { r = gguf_get_val_i16 (ctx_gguf, k); }
+    void gk_set(int k, int32_t     & r) { r = gguf_get_val_i32 (ctx_gguf, k); }
+    void gk_set(int k, int64_t     & r) { r = gguf_get_val_i64 (ctx_gguf, k); }
+    void gk_set(int k, float       & r) { r = gguf_get_val_f32 (ctx_gguf, k); }
+    void gk_set(int k, double      & r) { r = gguf_get_val_f64 (ctx_gguf, k); }
+    void gk_set(int k, bool        & r) { r = gguf_get_val_bool(ctx_gguf, k); }
+    void gk_set(int k, std::string & r) { r = std::string(gguf_get_val_str(ctx_gguf, k)); }
+
+    template<typename T>
+    typename std::enable_if<std::is_integral<T>::value, void>::type
+    gk_set(int k, struct gk_get_arrlen<T> & r) { r.output = gguf_get_arr_n(ctx_gguf, k); }
```

Comment:
Could you explain why you need a reference to a reference here? Could you remove the need for gk_get_arrlen with pointers?

Reply:
I was hoping you'd show up. :) I don't need a reference to the `gk_get_arrlen` struct, it could be passed by value. I think I do need it with the reference inside though, or at least I couldn't figure out a way to avoid that. This is basically my first attempt at using templates.

The point is to be able to replace something like:

```c++
GGUF_GET_KEY(ctx, hparams.n_vocab, gguf_get_arr_n, GGUF_TYPE_ARRAY, true, kv(LLM_KV_TOKENIZER_LIST));
```

The target here is a `uint32_t &` so I needed to wrap it in a container so the right template function could get selected in that case.

I think there are a few other places also where I'm needlessly passing stuff by reference.

Reply:
You could either give the function a different name, or disambiguate the overload using a dummy tag struct (similar to how [std::piecewise_construct](https://en.cppreference.com/w/cpp/utility/piecewise_construct) works).